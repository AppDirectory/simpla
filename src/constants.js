export const SERVER = 'https://api.simpla.io';
export const ELEMENTS_SERVER = 'https://elements.simpla.io';
export const APP_SERVER = 'https://app.simpla.io';

export const ELEMENTS = [
  'simpla-img/simpla-img.html',
  'simpla-text/simpla-text.html',
  'simpla-block/simpla-block.html',
  'sm-admin/sm-admin.html'
];

export const OPTIONS = {
  api: {
    auth: SERVER
  },
  elements: {
    base: `${ELEMENTS_SERVER}/`,
    paths: ELEMENTS
  }
}
