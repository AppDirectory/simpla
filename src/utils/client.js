import 'isomorphic-fetch';

/**
 * Check Status and request courtesy of feathers-rest
 * See https://github.com/feathersjs/feathers-rest/blob/master/src/client/fetch.js
 */
function checkStatus(response) {
  if (response.ok) {
    return response;
  }

  return Promise.resolve()
    .then(() => response.json())
    .then(error => {
      error.code = error.code || response.status;
      error.statusText = error.statusText || response.statusText;
      return Promise.reject(error);
    });
}

function request(options) {
  let fetchOptions = Object.assign({}, options),
      defaults = { Accept: 'application/json' };

  if (options.method !== 'GET') {
    defaults['Content-Type'] = 'application/json';
  }

  fetchOptions.headers = Object.assign({}, defaults, fetchOptions.headers);

  if (options.body) {
    fetchOptions.body = JSON.stringify(options.body);
  }

  return fetch(options.url, fetchOptions)
      .then(checkStatus)
      .then(response => response.json());
};

function requestWithToken(options) {
  let token = options.token;

  if (token) {
    options.headers = Object.assign({
      'Authorization': `Bearer ${token}`
    }, options.headers);
  }

  return request(options);
}

export default {
  get(url, options) {
    return request(Object.assign({ method: 'GET' }, options, { url }));
  },

  post(url, options) {
    return requestWithToken(Object.assign({ method: 'POST' }, options, { url }));
  },

  put(url, options) {
    return requestWithToken(Object.assign({ method: 'PUT' }, options, { url }));
  },

  delete(url, options) {
    return requestWithToken(Object.assign({ method: 'DELETE' }, options, { url }));
  }
}
