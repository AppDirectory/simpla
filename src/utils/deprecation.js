/**
 * NOTE: This is here purely for backwards compatibility, in future it should
 * 	be removed when possible
 * 	@return {undefined}
 */
export function supportDeprecatedInitializer(init) {
  let tag = document.querySelector('script[simpla-api]'),
      project = tag && tag.getAttribute('simpla-api');

  if (project) {
    console.warn(`The [simpla-api] attribute setup method is deprecated. Please use Simpla('${project}') instead, see https://www.simpla.io/docs/start`);
    init(project);
  }
}

/**
 * NOTE: Also for backwards compatibility. This below is a private API and can be
 * 	safely removed once all elements who depend on it are upgraded
 * 	@return {undefined}
 */
export function supportDeprecatedConfig(authEndpoint, project) {
  window.simpla = {
    config: {
      server: authEndpoint,
      api: project
    }
  };
}
