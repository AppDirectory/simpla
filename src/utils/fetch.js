import unfetch from 'unfetch';

// Doing this means that we're preferencing calling window.fetch always - which
//  means if window.fetch is altered in future, this will use it. This helps for
//  things like fetch-mock for mocking in tests
export default (...args) => window.fetch ? window.fetch(...args) : unfetch(...args);
