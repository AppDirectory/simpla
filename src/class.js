import { OPTIONS, SERVER } from './constants';
import { Simpla as SimplaCore } from 'simpla-core';

/**
 * Simpla class to boot the Simpla app.
 */
export default class Simpla extends SimplaCore {
  constructor(options, ...args) {

    if (typeof options === 'string') {
      options = { project: options };
    }

    // Add defaults
    options = Object.assign({}, OPTIONS, options);
    super(options, ...args);

    // Set project on options
    this.options.api.data = `${SERVER}/projects/${options.project}`;
  }

  /**
   * Load the given elements
   * @param  {Array}       hrefs hrefs for elements to load
   * @param  {String?}     base  Base path to prepend to elements, defaults to ''
   * @return {undefined}
   */
  static loadElements(hrefs = [], base = '') {
    hrefs
      .map(href => base + href)
      .forEach(href => {
        let link = document.createElement('link');
        link.href = href;
        link.rel = 'import';
        document.head.appendChild(link);
      });
  }
}
