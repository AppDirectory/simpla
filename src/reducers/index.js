import { DATA_PREFIX, QUERIES_PREFIX } from '../constants/state';
import { combineReducers } from 'redux';
import authenticated from './authenticated';
import options from './options';
import queries from './queries';
import imports from './imports';
import editable from './editable';
import data from './data';
import token from './token';
import save from './save';

const reducer = combineReducers({
  [ DATA_PREFIX ]: data,
  [ QUERIES_PREFIX ]: queries,
  _imports: imports,
  authenticated,
  config: options,
  editable,
  token,
  save
});

export default reducer;
