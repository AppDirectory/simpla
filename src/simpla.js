import 'es6-promise/auto';
import { createStore, applyMiddleware, compose } from 'redux';
import { setOption } from './actions/options';
import { editActive, editInactive } from './actions/editable';
import { login, logout } from './actions/authentication';
import { get, set, remove, find } from './actions/data';
import { observeQuery } from './actions/queries';
import save from './actions/save';
import { AUTH_SERVER } from './constants/options';
import { DATA_PREFIX, QUERIES_PREFIX, PUBLIC_STATES } from './constants/state';
import * as types from './constants/actionTypes';
import { configurePolymer } from './utils/prepare';
import {
  storeToObserver,
  dispatchThunkAndExpect,
  pathToUid,
  itemUidToPath,
  queryResultsToPath,
  validatePath,
  toQueryParams,
  uidsToResponse
} from './utils/helpers';
import ping from './plugins/ping';
import persistToken from './plugins/persistToken';
import thunk from 'redux-thunk';
import rootReducer from './reducers';

// Setup Polymer configuration
configurePolymer();

const Simpla = new class Simpla {
  constructor() {
    const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
    this._store = createStore(rootReducer, composeEnhancers(applyMiddleware(thunk)));
  }

  init(project) {
    this._store.dispatch(setOption('project', project));

    // Initialize endpoints
    this._store.dispatch(setOption('authEndpoint', AUTH_SERVER));
    this._store.dispatch(setOption('dataEndpoint', `${AUTH_SERVER}/projects/${project}/content`));
  }

  // Authentication
  login(...args) {
    return dispatchThunkAndExpect(this._store, login(...args), types.LOGIN_SUCCESSFUL);
  }

  logout(...args) {
    return dispatchThunkAndExpect(this._store, logout(...args), types.LOGOUT_SUCCESSFUL);
  }

  // Data
  find(options = {}) {
    let parentPath = options.parent;
    options.parent = pathToUid(parentPath);
    return Promise.resolve()
      .then(() => validatePath(parentPath))
      .then(() => dispatchThunkAndExpect(
        this._store,
        find(options),
        types.FIND_DATA_SUCCESSFUL
      ))
      .then(queryResultsToPath);
  }

  get(path, ...args) {
    const uid = pathToUid(path);
    return Promise.resolve()
      .then(() => validatePath(path))
      .then(() => dispatchThunkAndExpect(
        this._store,
        get(uid, ...args),
        types.GET_DATA_SUCCESSFUL
      ))
      .then(itemUidToPath);
  }

  set(path, ...args) {
    const uid = pathToUid(path);
    return Promise.resolve()
      .then(() => validatePath(path))
      .then(() => dispatchThunkAndExpect(
        this._store,
        set(uid, ...args),
        types.SET_DATA_SUCCESSFUL
      ))
      .then(itemUidToPath);
  }

  remove(path, ...args) {
    const uid = pathToUid(path);
    return Promise.resolve()
      .then(() => validatePath(path))
      .then(() => dispatchThunkAndExpect(
        this._store,
        remove(uid, ...args),
        types.REMOVE_DATA_SUCCESSFUL
      ))
      .then(itemUidToPath);
  }

  observe(path, ...args) {
    let callback = args.pop(),
        uid = pathToUid(path),
        pathInState,
        wrappedCallback;

    if (!uid) {
      throw new Error('Observe must be given a valid path');
    }

    validatePath(path);

    pathInState = [ DATA_PREFIX, 'content', uid ];
    wrappedCallback = () => this.get(path).then(callback);

    return storeToObserver(this._store).observe(pathInState, wrappedCallback);
  }

  observeQuery(query, callback) {
    let queryString,
        pathInStore,
        wrappedCallback;

    query.parent = pathToUid(query.parent);
    queryString = toQueryParams(query);
    pathInStore = [ QUERIES_PREFIX, queryString, 'matches' ];

    this._store.dispatch(observeQuery(query));

    wrappedCallback = (uids) => {
      return callback(
        queryResultsToPath(
          uidsToResponse(uids, this._store.getState())
        )
      );
    }

    return storeToObserver(this._store).observe(pathInStore, wrappedCallback);
  }

  save(...args) {
    return dispatchThunkAndExpect(this._store, save(...args), types.SAVE_SUCCESSFUL);
  }

  // Editable
  editable(on) {
    this._store.dispatch(on ? editActive() : editInactive());
  }

  // State
  getState(path) {
    let state = this._store.getState();

    if (path) {
      return PUBLIC_STATES.indexOf(path) === -1 ? undefined : state[path];
    }
    
    return PUBLIC_STATES.reduce((publicState, property) => {
      publicState[property] = state[property];
      return publicState;
    }, {});
  }

  observeState(...args) {
    return storeToObserver(this._store).observe(...args);
  }
}

// Init plugins
const plugins = [
  ping,
  persistToken
];

plugins.forEach(plugin => plugin(Simpla));

export default Simpla;
