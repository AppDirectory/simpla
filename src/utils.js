import { ELEMENTS_SERVER } from './constants';

/**
 * Ready web components in the browser. If webcomponents are natively supported
 * 	it will do nothing, otherwise it will load in the webcomponents polyfills
 * @return {Promise}  Promise which resolves once web components is definitely ready
 */
export function readyWebComponents() {
  // Conditionally load WCjs
  const webComponentsSupported =
      'registerElement' in document &&
      'import' in document.createElement('link') &&
      'content' in document.createElement('template');

  if (webComponentsSupported) {
    return Promise.resolve();
  }

  /**
   * Load in web components polyfill
   * @return {Promise} Promise which resolves once finished
   */
  return new Promise(resolve => {
    let script = document.createElement('script');
    script.async = true;
    script.src = `${ELEMENTS_SERVER}/webcomponentsjs/webcomponents-lite.min.js`;
    script.onload = resolve;
    document.head.appendChild(script);
  });
};

/**
 * Configure Polymer with dom = shadow, unless Polymer is already defined
 * @return {undefined}
 */
export function configurePolymer() {
  window.Polymer = window.Polymer || { dom: 'shadow' };
}
