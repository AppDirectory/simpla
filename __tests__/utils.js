import * as utils from '../src/utils';

describe('utilities', () => {
  beforeAll(() => {
    // To deal with randomString()
    Math._random = Math.random;
    Math.random = jest.fn(() => 0.55);
  });

  afterAll(() => {
    Math.random = Math._random;
  });

  test('isPrimitive', () => {
    const cases = [
      {
        value: true,
        primitive: true
      },
      {
        value: false,
        primitive: true
      },
      {
        value: null,
        primitive: true
      },
      {
        value: undefined, // eslint-disable-line no-undefined
        primitive: true
      },
      {
        value: 'foo',
        primitive: true
      },
      {
        value: 0,
        primitive: true
      },
      {
        value: {},
        primitive: false
      },
      {
        value: [],
        primitive: false
      }
    ];

    cases.forEach(({ value, primitive }) => {
      expect(utils.isPrimitive(value)).toBe(primitive);
    });
  });

  test('deepMapIn', () => {
    const mapOver = {
      num: 4,
      str: ['foo'],
      arr: [
        4,
        'bar',
        {
          str: 'foo'
        }
      ],
      obj: {
        num: 6,
        str: 'bar'
      },
      bool: true
    };

    const reverseAndNegate = value => {
      if (typeof value === 'number') {
        return -value;
      }

      if (typeof value === 'string') {
        return value.split('').reverse().join('');
      }

      return value;
    };

    expect(utils.deepMapIn(mapOver, reverseAndNegate)).toMatchObject({
      num: -4,
      str: ['oof'],
      arr: [
        -4,
        'rab',
        {
          str: 'oof'
        }
      ],
      obj: {
        num: -6,
        str: 'rab'
      },
      bool: true
    });
  });

  test('toContentAndUploads;', () => {
    const entries = [
      {
        path: '/image',
        data: {
          src:
            'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7'
        }
      },
      {
        path: '/article',
        data: {
          html: `
            <p>Hello World</p>
            <img src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7">
            `
        }
      }
    ];

    expect(
      entries.map(utils.toShortcodeAndUploads)
    ).toMatchSnapshot();
  });
});
