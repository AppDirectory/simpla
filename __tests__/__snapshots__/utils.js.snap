// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utilities toContentAndUploads; 1`] = `
Array [
  Array [
    Object {
      "data": Object {
        "src": "[upload 6cpj8.gif]",
      },
      "path": "/image",
    },
    Array [
      Array [
        "6cpj8.gif",
        "R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7",
      ],
    ],
  ],
  Array [
    Object {
      "data": Object {
        "html": "
            <p>Hello World</p>
            <img src=\\"[upload 6cpj8.gif]\\">
            ",
      },
      "path": "/article",
    },
    Array [
      Array [
        "6cpj8.gif",
        "R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7",
      ],
    ],
  ],
]
`;
